### DISCUSSION TITLE
Communication box spawner

### DISCUSSION THREAD

**Discussion by tiagocunico** [Open-Industry-Project/Open-Industry-Project#79] [Ideas]
It would be nice if implemented the communication on the box spawner too, so we can control if it will spawn and the spawn interval.

**Comment by ryevdokimov** [comment]
I'm not sure it makes sense to use a controller like a PLC to manage things like spawners. Can you provide an example of a use case? It might be an interesting idea to provide a mechanism to allow users to easily externally expose whatever they want through the editor just by modifying some predefined properties. 

In the meantime, it is possible to modify the properties of the box spawner through scripting. Godot's built in scripting language GDScript is pretty simple and is tightly integrated with the editor to allow things like dragging and dropping node references into the text editor. From there you can modify the spawn interval. 

Here is a very simple script attached to a BoxSpawnerManager node I created, that slowly increases the spawn interval over time.

![image](https://github.com/user-attachments/assets/86e3a507-621a-41d6-a24b-1d4c0a40811a)


Including a property to enable/disable spawning is a good suggestion.

**Reply by tiagocunico** [reply]
Thanks for the tip with GDScript, I'm very new with Godot but very excited to learn and contribute :). 

I work a lot with the food industry and some machines I need to send a signal to tell them to stop sending boxes ahead because the line in front of them is full. Today to simulate this I'm using a chain transfer that send the box out of the conveyor and to a despawner. It would be easier to disable the spawn with a tag.

**Reply by ryevdokimov** [reply]
>I need to send a signal to tell them to stop sending boxes ahead because the line in front of them is full.

I'm assuming you're signaling personnel with a beacon of some sort. You could achieve this now by adding a beacon to the scene and using GDScript to tie it together with the BoxSpawner. Now if you trigger an output in the PLC to drive the beacon light, it will also affect the box spawner via GDScript. 

![image](https://github.com/user-attachments/assets/0d757485-069f-45a9-bce8-7fcc23b131a2)

**Reply by tiagocunico** [reply]
We end up not signaling with a beacon because the line stops and run a lot of times during the day, so it's normal to them the line stops, until they manage to palletize some of the boxes and the line can run again.

I like the workaround with the beacon, I will implement it. Thanks

**Reply by ryevdokimov** [reply]
>Including a property to enable/disable spawning is a good suggestion

This was implemented: be055dc200da37b09fec3d35ebd79ed263788623

**Comment by fchaxel** [comment]
Hi @ryevdokimov 

You've say 

> I'm not sure it makes sense to use a controller like a PLC to manage things like spawners. Can you provide an example of a use case?

I'm OK with @tiagocunico, it should be, "requested" in issue #245 

An example ? Yes. I've been a Nestle Waters one week ago (Perrier, Vitel, Contrex), but I have a lot of more real exemples.
One botteling machine, and a long, long ... long queue of conveyors, then a packaging machine.

When the queue is full (the packaging is too slow, or in error, ...) the botteling machine stop it's work because a sensor give it the information and/or because the packaging machine send network message for that (or stop sending periodic ack messages).

Let say we are modeling the packaging machine with OIP. The bottling machine is the spawner, and it cannot be stopped today. With a simple modification it can be.

... I've just see the Conveyor property ...(Spawning pauses when conveyor speed is zero) it's a possible solution, but it do not change my mind.

With such an input from OIPComs, one can simple imagine also to use for instance the Opc Ua server Kepware with the simulation driver to connect an infrared OIP sensor to a tag, then connect also this Tag to the validation of the spawner. The spawner has a simple closed-loop control bit with it's associated sensor.

Adds : one can also imagine an internal OIPCom driver named internal for that and more ... I can develop it, say me.

OIPComs validation can be done with the both spawners.

Bye.
